### PR TITLE
WIP: Handle cancelling the OCI functions on exceeding user context deadline

### PR DESCRIPTION
**Pull request by sudarshan12s**
- Issue
currently Few OCI calls like ocistmtfetch2 ( called from rows.Next), LOB functions OCILobCreateTemporary (called from dataSetLOB) are not cancelled either through break or calltimeout in case they are exceeding context deadline. 
The PR tries to address this by calling a wrapper function which is called at beginning of QueryContext and ExecContext.

- Following changes are done
-- wrapper function, handleDeadline is added  to handle context deadline for ODPI C functions. The function calls uses OCIBreak on exceeding deadline.
--This function is kept in QueryContext/ExecContext functions to handle all the ODPI calls that subsequently get called but not repeated before every ODPI call.   The wrapper , handleDeadline spawns a  go routine and on reaching deadline calls OCIBreak. 
-- rows.Next doesnt take context parameter, the context parameter is saved in stmt and is fetched in rows.Next. handle null case of context. 
-- OCIStmtPrepare2 doesnt do roundtrip, hence removed calling OCIBreak. sorry for incorrect infn earlier

- Testing:
TestStmtFetchDeadlineForLOB is added in z_test.go. It basically cancels the OCIStmtFetch2 call. 

- TestStmtFetchDeadlineForLOB
=== RUN   TestStmtFetchDeadlineForLOB
    z_test.go:3779: Info: Next: ORA-00000: DPI-1067: call timeout of 4980 ms exceeded with ORA-3156
--- PASS: TestStmtFetchDeadlineForLOB (5.17s)
PASS
ok           github.com/godror/godror          5.601s

 
ODPI logs 
ODPI [22243] 2020-10-12 03:29:45.426: fn start dpiStmt_fetchRows(0x145d960)
ODPI [22261] 2020-10-12 03:29:50.159: fn start dpiConn_breakExecution(0x1454540)
ODPI [22261] 2020-10-12 03:29:50.161: fn end dpiConn_breakExecution(0x1454540) -> 0
ODPI [22243] 2020-10-12 03:29:51.086: fn end dpiStmt_fetchRows(0x145d960) -> -1
 
- TestIssue100
=== RUN   TestIssue100
    z_test.go:3669: error: dpiStmt_execute: context deadline exceeded
    z_test.go:3675: Error: Timeout
--- PASS: TestIssue100 (2.12s)
PASS
ok           github.com/godror/godror          2.498s
 
 
 
ODPI logs 
ODPI [22776] 2020-10-12 03:32:56.410: fn start dpiStmt_execute(0x298a480)
ODPI [22778] 2020-10-12 03:32:58.408: fn start dpiConn_breakExecution(0x2987c60)
ODPI [22778] 2020-10-12 03:32:58.410: fn end dpiConn_breakExecution(0x2987c60) -> 0
ODPI [22776] 2020-10-12 03:32:59.094: fn end dpiStmt_execute(0x298a480) -> -1

- Tested with sample program which runs for 1000 iterations with single connection in session pool. Each iteration does a queryContext with smaller timeout which would fail and then run with higher timeout. It is done to test OCIBreak multiple times in a single connection and re-using it for subsequent queries . 

- Basic Sanity

Go tests are run with StandAlone=0/1.

--- PASS: TestMaybeBadConn (0.00s)
--- PASS: TestCalculateTZ (0.02s)
--- PASS: TestDataSetGet (0.00s)
--- PASS: TestFromErrorInfo (0.00s)
--- PASS: TestMarshalJSON (0.00s)
--- PASS: TestMapToSlice (0.00s)
--- PASS: TestContextWithUserPassw (0.10s)
--- PASS: TestPlSqlObjectDirect (0.31s)
--- PASS: TestOpenBadMemory (0.06s)
--- PASS: TestRanaOraIssue244 (0.60s)
--- PASS: TestExecHang (1.01s)
--- PASS: TestSDO (0.29s)
--- PASS: TestIssue134 (0.20s)
 
=== RUN   TestPreFetchQuery
    z_test.go:3554: SingleRowFetch:  Prefetch: -99 , Arraysize: -99 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: -99 , Arraysize: -99 , Rows:  214 , Round-trips: 4
    z_test.go:3554: SingleRowFetch:  Prefetch: 0 , Arraysize: -99 , Rows:  1 , Round-trips: 2
    z_test.go:3560: MultiRowFetch:  Prefetch: 0 , Arraysize: -99 , Rows:  214 , Round-trips: 4
    z_test.go:3554: SingleRowFetch:  Prefetch: 1 , Arraysize: -99 , Rows:  1 , Round-trips: 2
    z_test.go:3560: MultiRowFetch:  Prefetch: 1 , Arraysize: -99 , Rows:  214 , Round-trips: 4
    z_test.go:3554: SingleRowFetch:  Prefetch: 2 , Arraysize: -99 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: 2 , Arraysize: -99 , Rows:  214 , Round-trips: 4
    z_test.go:3554: SingleRowFetch:  Prefetch: 100 , Arraysize: -99 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: 100 , Arraysize: -99 , Rows:  214 , Round-trips: 3
    z_test.go:3554: SingleRowFetch:  Prefetch: -1 , Arraysize: 100 , Rows:  1 , Round-trips: 2
    z_test.go:3560: MultiRowFetch:  Prefetch: -1 , Arraysize: 100 , Rows:  214 , Round-trips: 4
testDb: busy=2 open=4 max=16 maxLifetime=5m0s timeout=1m0s waitTimeout=5s <nil>
    z_test.go:3554: SingleRowFetch:  Prefetch: 0 , Arraysize: 100 , Rows:  1 , Round-trips: 2
    z_test.go:3560: MultiRowFetch:  Prefetch: 0 , Arraysize: 100 , Rows:  214 , Round-trips: 4
    z_test.go:3554: SingleRowFetch:  Prefetch: 1 , Arraysize: 100 , Rows:  1 , Round-trips: 2
    z_test.go:3560: MultiRowFetch:  Prefetch: 1 , Arraysize: 100 , Rows:  214 , Round-trips: 4
    z_test.go:3554: SingleRowFetch:  Prefetch: 2 , Arraysize: 100 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: 2 , Arraysize: 100 , Rows:  214 , Round-trips: 4
   z_test.go:3554: SingleRowFetch:  Prefetch: 100 , Arraysize: 100 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: 100 , Arraysize: 100 , Rows:  214 , Round-trips: 3
    z_test.go:3554: SingleRowFetch:  Prefetch: -99 , Arraysize: 40 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: -99 , Arraysize: 40 , Rows:  214 , Round-trips: 7
    z_test.go:3554: SingleRowFetch:  Prefetch: 2 , Arraysize: 40 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: 2 , Arraysize: 40 , Rows:  214 , Round-trips: 7
    z_test.go:3554: SingleRowFetch:  Prefetch: -1 , Arraysize: 40 , Rows:  1 , Round-trips: 2
    z_test.go:3560: MultiRowFetch:  Prefetch: -1 , Arraysize: 40 , Rows:  214 , Round-trips: 7
    z_test.go:3554: SingleRowFetch:  Prefetch: 120 , Arraysize: -99 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: 120 , Arraysize: -99 , Rows:  214 , Round-trips: 3
    z_test.go:3554: SingleRowFetch:  Prefetch: 120 , Arraysize: 100 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: 120 , Arraysize: 100 , Rows:  214 , Round-trips: 3
testDb: busy=2 open=2 max=16 maxLifetime=5m0s timeout=1m0s waitTimeout=5s <nil>
    z_test.go:3554: SingleRowFetch:  Prefetch: 120 , Arraysize: 0 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: 120 , Arraysize: 0 , Rows:  214 , Round-trips: 3
    z_test.go:3554: SingleRowFetch:  Prefetch: 120 , Arraysize: -1 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: 120 , Arraysize: -1 , Rows:  214 , Round-trips: 3
    z_test.go:3554: SingleRowFetch:  Prefetch: 120 , Arraysize: 250 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: 120 , Arraysize: 250 , Rows:  214 , Round-trips: 2
    z_test.go:3554: SingleRowFetch:  Prefetch: 0 , Arraysize: 10 , Rows:  1 , Round-trips: 2
    z_test.go:3560: MultiRowFetch:  Prefetch: 0 , Arraysize: 10 , Rows:  214 , Round-trips: 23
    z_test.go:3554: SingleRowFetch:  Prefetch: 10 , Arraysize: 10 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: 10 , Arraysize: 10 , Rows:  214 , Round-trips: 22
    z_test.go:3554: SingleRowFetch:  Prefetch: 214 , Arraysize: 214 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: 214 , Arraysize: 214 , Rows:  214 , Round-trips: 2
    z_test.go:3554: SingleRowFetch:  Prefetch: 215 , Arraysize: 214 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: 215 , Arraysize: 214 , Rows:  214 , Round-trips: 1
    z_test.go:3554: SingleRowFetch:  Prefetch: 215 , Arraysize: -99 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: 215 , Arraysize: -99 , Rows:  214 , Round-trips: 1
testDb: busy=1 open=2 max=16 maxLifetime=5m0s timeout=1m0s waitTimeout=5s <nil>
    z_test.go:3554: SingleRowFetch:  Prefetch: 215 , Arraysize: 10 , Rows:  1 , Round-trips: 1
    z_test.go:3560: MultiRowFetch:  Prefetch: 215 , Arraysize: 10 , Rows:  214 , Round-trips: 1
--- PASS: TestPreFetchQuery (91.15s)
PASS

OCI calls roundtrips doc:
https://docs.oracle.com/en/database/oracle/oracle-database/19/lnoci/oci-function-server-round-trips.html#GUID-9B2F05F9-D841-4493-A42D-A7D89694A2D1

### CODE REVIEW

**Comment by tgulacsi** [comment]
Regarding dpiConn_setCallTimeout: it was removed in 4d9c582561c1391a8cfd1cff8ad78d65916d3aa6 as it may leave the underlying connection in an unusable state, which may or may not be reset.

But that may were 'caused by the dpiStmt_execute on a different thread (goroutine).
So I'm not against reintroducing setCallTimeout, but please test thoroughly.

File: stmt.go [hunk 1]
```diff
@@ -335,13 +336,15 @@ func (st *statement) ExecContext(ctx context.Context, args []driver.NamedValue)
...
 	}
+
 	Log := ctxGetLog(ctx)
 
 	st.Lock()
 	defer st.Unlock()
 	if st.conn == nil {
 		return nil, driver.ErrBadConn
 	}
+	st.ctx = ctx
```
Comment (tgulacsi) [inline stmt.go:347]: Clear st.ctx in st.closeNotLocked.

**Review by tgulacsi — CHANGES_REQUESTED** [review]

File: stmt.go (hunk 1 above)
Comment (sudarshan12s) [reply stmt.go:348]: reset context to context.Background . ok?

**Comment by sudarshan12s** [comment]
> Regarding dpiConn_setCallTimeout: it was removed in [4d9c582](https://github.com/godror/godror/commit/4d9c582561c1391a8cfd1cff8ad78d65916d3aa6) as it may leave the underlying connection in an unusable state, which may or may not be reset.
> 
> But that may were 'caused by the dpiStmt_execute on a different thread (goroutine).
> So I'm not against reintroducing setCallTimeout, but please test thoroughly.

yes testing in progress, shall update.

File: stmt.go (hunk 1 above)
Comment (tgulacsi) [reply stmt.go:348]: A nil would be better.

**Comment by sudarshan12s** [comment]
> Regarding dpiConn_setCallTimeout: it was removed in [4d9c582](https://github.com/godror/godror/commit/4d9c582561c1391a8cfd1cff8ad78d65916d3aa6) as it may leave the underlying connection in an unusable state, which may or may not be reset.
> 
> But that may were 'caused by the dpiStmt_execute on a different thread (goroutine).
> So I'm not against reintroducing setCallTimeout, but please test thoroughly.

I have updated description and changes to keep only OCIBreak. have not seen issues with calltimeout though.  After some discussion/thoughts, OCIBreak serves the purpose , hence retained the same. Please see.

**Comment by sudarshan12s** [comment]
> > Regarding dpiConn_setCallTimeout: it was removed in [4d9c582](https://github.com/godror/godror/commit/4d9c582561c1391a8cfd1cff8ad78d65916d3aa6) as it may leave the underlying connection in an unusable state, which may or may not be reset.
> > But that may were 'caused by the dpiStmt_execute on a different thread (goroutine).
> > So I'm not against reintroducing setCallTimeout, but please test thoroughly.
> 
> yes testing in progress, shall update.

I have updated description and changes to keep only OCIBreak. have not seen issues with calltimeout though. After some discussion/thoughts, OCIBreak serves the purpose , hence retained the same. Please see.

File: conn.go [hunk 2]
```diff
@@ -87,6 +123,7 @@ func (c *conn) Break() error {
 		return nil
 	}
 	if C.dpiConn_breakExecution(c.dpiConn) == C.DPI_FAILURE {
+		Log("msg", "Break", "error", c.getError())
```
Comment (tgulacsi) [inline conn.go:126]: Log may be nil, so please guard it with "if Log != nil"

File: stmt.go [hunk 3]
```diff
@@ -359,6 +363,13 @@ func (st *statement) ExecContext(ctx context.Context, args []driver.NamedValue)
 		st.closeNotLocking()
 		return maybeBadConn(err, c)
 	}
+    
+    // HandleDeadline for all ODPI calls called below
+	done := make(chan struct{})
+    defer close(done)
```
Comment (tgulacsi) [inline stmt.go:369]: Please `go fmt`

**Review by tgulacsi — CHANGES_REQUESTED** [review]

File: conn.go (hunk 2 above)
Comment (sudarshan12s) [reply conn.go]: Thanks . done

File: stmt.go (hunk 3 above)
Comment (sudarshan12s) [reply stmt.go]: done.

**Review by tgulacsi — APPROVED** [review]

**Pull request merged**